pkg/frontend/querymiddleware/cardinality.go

```diff
@@ -0,0 +1,155 @@
+
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+const (
+	// cardinalityEstimateTTL is how long a key must see no write to expire and be removed from the cache.
+	cardinalityEstimateTTL = 7 * 24 * time.Hour
+
+	// cardinalityEstimateBucketSize is the size of buckets that queries are bucketed into.
+	cardinalityEstimateBucketSize = 24 * time.Hour
+
+	// cacheErrorToleranceFraction is how much the estimate must deviate
+	// from the actually observed cardinality to update the cache.
+	cacheErrorToleranceFraction = 0.1
+)
+
+type cardinalityEstimationMetrics struct {
+	estimationError prometheus.Histogram
+}
+
+// cardinalityEstimation is a Handler that caches estimates for a query's
+// cardinality based on similar queries seen previously.
+type cardinalityEstimation struct {
+	cache  cache.Cache
+	next   Handler
+	logger log.Logger
+
+	cardinalityEstimationMetrics
+}
+
+func newCardinalityEstimationMiddleware(cache cache.Cache, logger log.Logger, registerer prometheus.Registerer) Middleware {
+	metrics := cardinalityEstimationMetrics{estimationError: promauto.With(registerer).NewHistogram(prometheus.HistogramOpts{
+		Name:    "cortex_query_frontend_cardinality_estimation_difference",
+		Help:    "Difference between estimated and actual query cardinality",
+		Buckets: prometheus.ExponentialBuckets(10, 10, 6),
+	})}
+	return MiddlewareFunc(func(next Handler) Handler {
+		return &cardinalityEstimation{
+			cache:  cache,
+			next:   next,
+			logger: logger,
+
+			cardinalityEstimationMetrics: metrics,
+		}
+	})
+}
+
+// Do injects a cardinality estimate into the query hints (if available) and
+// caches the actual cardinality observed for this query.
+func (c *cardinalityEstimation) Do(ctx context.Context, request Request) (Response, error) {
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, c.logger, "cardinalityEstimation.Do")
+	defer spanLog.Finish()
+
+	tenants, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return c.next.Do(ctx, request)
+	}
+
+	k := generateCardinalityEstimationCacheKey(tenant.JoinTenantIDs(tenants), request, cardinalityEstimateBucketSize)
+
+	var estimatedCardinality uint64
+	if estimate, ok := c.lookupCardinalityForKey(ctx, k); ok {
+		estimatedCardinality = estimate
+		request = request.WithEstimatedCardinalityHint(estimate)
+	}
+
+	res, err := c.next.Do(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	statistics := stats.FromContext(ctx)
+	actualCardinality := statistics.GetFetchedSeriesCount()
+
+	if !isCardinalitySimilar(actualCardinality, estimatedCardinality) {
+		c.storeCardinalityForKey(ctx, k, actualCardinality)
+	}
+
+	if estimatedCardinality > 0 {
+		c.estimationError.Observe(math.Abs(float64(actualCardinality) - float64(estimatedCardinality)))
+		statistics.AddEstimatedSeriesCount(estimatedCardinality)
+		spanLog.LogKV("estimated cardinality", estimatedCardinality, "actual cardinality", actualCardinality)
```

Comment:
A couple of nits:
1. Use LogFields() instead of LogKV() (see their doc, LogFields() is recommended)
2. remove estimated cardinality from here cause moved above (following another suggestion I left)